pulsar-broker/src/main/java/org/apache/pulsar/broker/service/HashRangeStickyKeyConsumerSelector.java

```diff
@@ -0,0 +1,109 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.broker.service;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.ConcurrentSkipListMap;
+
+/**
+ * This is a consumer selector based hash range.
+ *
+ * 1.Each consumer serves a fixed range of hash value
+ * 2.The whole range of hash value could be covered by all the consumers.
+ * 3.Once a consumer is removed, the left consumers could still serve the whole range.
+ */
+public class HashRangeStickyKeyConsumerSelector implements StickyKeyConsumerSelector {
+
+    public static final int RANGE_SIZE = 2 << 15;
+
+    private final ConcurrentSkipListMap<Integer, Consumer> rangeMap;
+    private final Map<Consumer, Integer> consumerRange;
+
+    public HashRangeStickyKeyConsumerSelector() {
+        this.rangeMap = new ConcurrentSkipListMap<>();
+        this.consumerRange = new HashMap<>();
+    }
+
+    @Override
+    public synchronized void addConsumer(Consumer consumer) {
+        if (rangeMap.size() == 0) {
+            rangeMap.put(RANGE_SIZE, consumer);
+            consumerRange.put(consumer, RANGE_SIZE);
+        } else {
+            splitRange(findBiggestRange(), consumer);
+        }
+    }
+
+    @Override
+    public synchronized void removeConsumer(Consumer consumer) {
+        Integer removeRange = consumerRange.get(consumer);
+        if (removeRange != null) {
+            if (removeRange == RANGE_SIZE && rangeMap.size() > 1) {
+                Consumer lowerConsumer = rangeMap.lowerEntry(removeRange).getValue();
+                rangeMap.put(removeRange, lowerConsumer);
+                consumerRange.put(lowerConsumer, removeRange);
+            } else {
+                rangeMap.remove(removeRange);
+                consumerRange.remove(consumer);
+            }
+        }
+    }
+
+    @Override
+    public Consumer select(String stickyKey) {
+        if (rangeMap.size() > 0) {
+            int slot = Math.abs(stickyKey.hashCode() % RANGE_SIZE);
```

Comment:
I would avoid using `hashCode` here. Use `Murmur3_32Hash`. 

You can move `Hash` interface and its implementations (`Murmur3_32Hash`) to pulsar-common. since it can be used both for client and broker.

Reply:
👍